drivers/can/can_mcp251xfd.c

```diff
@@ -0,0 +1,1814 @@
+			   BIT(dev_cfg->int_gpio_dt.pin));
+
+	if (gpio_add_callback(dev_cfg->int_gpio_dt.port, &(dev_data->int_gpio_cb)) < 0) {
+		return -EINVAL;
+	}
+
+	if (gpio_pin_interrupt_configure_dt(&dev_cfg->int_gpio_dt, GPIO_INT_LEVEL_ACTIVE) < 0) {
+		return -EINVAL;
+	}
+
+	k_thread_create(&dev_data->int_thread, dev_data->int_thread_stack,
+			CONFIG_CAN_MCP251XFD_INT_THREAD_STACK_SIZE,
+			(k_thread_entry_t)mcp251xfd_int_thread, (void *)dev, NULL, NULL,
+			K_PRIO_COOP(CONFIG_CAN_MCP251XFD_INT_THREAD_PRIO), 0, K_NO_WAIT);
+
+	(void)k_thread_name_set(&dev_data->int_thread, "MCP251XFD interrupt thread");
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	ret = mcp251xfd_reset(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to reset the device [%d]", ret);
+		goto done;
+	}
+
+	ret = mcp251xfd_init_timing_struct(&timing, dev, &dev_cfg->timing_params, true);
+	if (ret < 0) {
+		LOG_ERR("Can't find timing for given param");
+		goto done;
+	}
+
+#if defined(CONFIG_CAN_FD_MODE)
+	ret = mcp251xfd_init_timing_struct(&timing_data, dev, &dev_cfg->timing_params_data, false);
+	if (ret < 0) {
+		LOG_ERR("Can't find timing for given param");
+		goto done;
+	}
+#endif
+
+	reg = mcp251xfd_read_crc(dev, MCP251XFD_REG_CON, MCP251XFD_REG_SIZE);
+	if (!reg) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	*reg = sys_le32_to_cpu(*reg);
+
+	opmod = FIELD_GET(MCP251XFD_REG_CON_OPMOD_MASK, *reg);
+
+	if (opmod != MCP251XFD_REG_CON_MODE_CONFIG) {
+		LOG_ERR("Device did not reset into configuration mode [%d]", opmod);
+		ret = -EIO;
+		goto done;
+	}
+
+	dev_data->current_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_CONFIG;
+
+	ret = mcp251xfd_init_con_reg(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_init_osc_reg(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_init_iocon_reg(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_init_int_reg(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_set_tdc(dev, false, 0);
+	if (ret < 0) {
+		goto done;
+	}
+
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	ret = mcp251xfd_init_tscon(dev);
+	if (ret < 0) {
+		goto done;
+	}
+#endif
+
+	ret = mcp251xfd_init_tef_fifo(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_init_tx_queue(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = mcp251xfd_init_rx_fifo(dev);
+	if (ret < 0) {
+		goto done;
+	}
+
+	LOG_DBG("%d TX FIFOS: 1 element", MCP251XFD_TX_QUEUE_ITEMS);
+	LOG_DBG("1 RX FIFO: %d elements", MCP251XFD_RX_FIFO_ITEMS);
+	LOG_DBG("%db of %db RAM Allocated",
+		MCP251XFD_TEF_FIFO_SIZE + MCP251XFD_TX_QUEUE_SIZE + MCP251XFD_RX_FIFO_SIZE,
+		MCP251XFD_RAM_SIZE);
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = can_set_timing(dev, &timing);
+	if (ret < 0) {
+		return ret;
+	}
+
+#if defined(CONFIG_CAN_FD_MODE)
+	ret = can_set_timing_data(dev, &timing_data);
+	if (ret < 0) {
+		return ret;
+	}
+#endif
+
+#if defined(CONFIG_CAN_FD_MODE)
+	ret = can_set_mode(dev, CAN_MODE_FD);
```

Comment:
All CAN controllers need to be in normal mode after boot up, regardless of `CONFIG_CAN_FD_MODE`.